### DISCUSSION TITLE
New to ZTNET - Full selfhosted wish but...

### DISCUSSION THREAD

**Discussion by ViP3R76** [sinamics/ztnet#466] [Q&A]
Hi, i am new to ZTNET and it is a wonderful gold piece of software. But, for now i can not achieve everything i want.

- VPS Server running as fully enclosed system
- every connection should be in the ZT network (excluded - ssh access to the VPS with external IP)
- every service running on the VPS should be only reachable by the ZTnet

My problems:
- i have read nearly all discussions here, and the point is to have 2 seperate instances (one for controller, one for the "client")
- for easier management i use portainer on the VPS Host-System, which exposes its ports only to the ZTNet (this works, if i only use the regular ZT connection - as it is bound to a ZT eth "device")
- i have around 5-6 external PCs that run in that ZT-Network - but only for connection inside this net - no dns (this is still done outside the ztnet with their local networks)
- i created a new planet (for selfhosting) and the connections are working as intended after replacing the planet.custom->planet within the normal windows ZT Client software and connection is done properly and registering to the controller

My main question:
- how can i achieve, that my only VPS is the controller AND can serve the needed ZT-device to itself?
- excluded options: 2nd VPS, no proxmox
- is it an option to have the controller on port 9994 (as i have a new planet and need to customize the connection to the controller in every case) and the local ZT Client on 9993 ?

**Comment by PizzaProgram** [comment]
> how can i achieve, that my only VPS is the controller AND can serve the needed ZT-device to itself?

This took me a while too, to realize, how easy is to do. (Maybe that's why it is not in the docs.):

Simply join the network from inside the docker container! 
`docker exec zerotier zerotier-cli join **{net-id}**`
I recommend adding a xxx.xxx.xxx.001 IP to it, before accepting on the admin page.

To add services to it, I think You have 2 options:
1. extend the docker stack
2. create a separate docker container and join the network of `zerotier`. 
![kép](https://github.com/user-attachments/assets/cf21d4c5-1bcb-4326-b6df-a4956d6664a1)

**Comment by ViP3R76** [comment]
Hm, this would be brilliant. This would be too easy. Will try it on the weekend :-) Thanks for the hint anyways!
